scripts/cfengine-build-host-setup.cf

```diff
@@ -0,0 +1,188 @@
+    redhat_8|centos_8::
+      "have_fakeroot" expression => returnszero("command -v fakeroot >/dev/null", "useshell");
+
+  commands:
+    !have_opt_jdk21.(debian_9|ubuntu_16)::
+      "sh $(this.promise_dirname)/linux-install-jdk21.sh" contain => in_shell;
+    (redhat_7|centos_7|redhat_8|centos_8|redhat_9).!have_development_tools::
+      "yum groups install 'Development Tools'" contain => in_shell;
+    (redhat_8|centos_8).!have_fakeroot:: # special fakeroot, missing from _8 an d up?
+      "sudo rpm -iv https://kojipkgs.fedoraproject.org//packages/fakeroot/1.23/1.fc29/x86_64/fakeroot-1.23-1.fc29.x86_64.rpm https://kojipkgs.fedoraproject.org//packages/fakeroot/1.23/1.fc29/x86_64/fakeroot-libs-1.23-1.fc29.x86_64.rpm"
+        contain => in_shell;
+
+
+  classes:
+    debian_11::
+      "have_pip2" expression => fileexists("/usr/local/bin/pip");
+    ubuntu_16::
+      "have_i386_architecture" expression => strcmp(execresult("${paths.dpkg} --print-foreign-architectures", "noshell"), "i386");
+    ubuntu::
+      "have_localhost_hostname" expression => strcmp(execresult("${paths.hostname}", "noshell"), "localhost.localdomain");
+    opensuse|suse|sles::
+      "have_$(suse_users_and_groups)_group" expression => returnszero("grep $(suse_users_and_groups) /etc/group >/dev/null", "useshell");
+      "have_$(suse_users_and_groups)_user" expression => returnszero("grep '^$(suse_users_and_groups):' /etc/passwd >/dev/null", "useshell");
+
+  files:
+    debian_9::
+      "/etc/apt/sources.list.d/*"
+        delete => tidy;
+      "/etc/apt/sources.list"
+        content => "deb http://archive.debian.org/debian/ stretch main contrib non-free";
+    suse|opensuse|sles::
+      # ensure %dist works in RPM .spec files - needed to add OS name/version to rpm filename
+      "/home/jenkins/.rpmmacros" content => "%dist .suse15";
+
+  commands:
+    redhat_9::
+      "sudo yum --nobest -e 0 -d 0 -y update" contain => in_shell; # 2022-11-15 rhel-9 needed --nobest in order to complete update/upgrade
+    redhat_8|redhat_9::
+      "sudo sed -ri 's/^%_enable_debug_packages/#\0/' /usr/lib/rpm/redhat/macros" contain => in_shell;
+    debian_11.!have_pip2::
+      "wget https://bootstrap.pypa.io/pip/2.7/get-pip.py -O get-pip.py && python2 get-pip.py && pip install psycopg2-binary"
+        contain => in_shell;
+
+    ubuntu_16.!have_i386_architecture:: # mingw build host
+      "${paths.dpkg} --add-architecture i386";
+
+    ubuntu.!have_localhost_hostname:: # hack for aws ubuntu hosts having unique ip-n-n-n-n hostnames, we need localhost.localdomain
+      "/usr/bin/hostnamectl set-hostname localhost.localdomain";
+    !have_daemon_group.(suse|sles|opensuse)::
+      "groupadd -g 1 daemon" contain => in_shell;
+    !have_bin_group.(suse|sles|opensuse)::
+      "groupadd -g 2 bin" contain => in_shell;
+    !have_sys_group.(suse|sles|opensuse)::
+      "groupadd -g 3 sys" contain => in_shell;
+    !have_daemon_user.(suse|sles|opensuse)::
```

Comment:
@nickanderson  I know that here I could use users promises but I liked the symmetry of doing it the same way for all these bits. Plus we don't have groups first-class and if a custom promise module then my trick of using mpf minimally would break down I think. FWIW.

Reply:
Yeah, understood and agree.